### ISSUE TITLE
How-to use MongoDB index directly for querying

### ISSUE THREAD

**Issue by psh-tc** [Stratio/Spark-MongoDB#72]
Hi there,

I have a MongoDB collection that contains time series data. It contains several million entries. There is an index on the "createdAt"-Field. I want to select a subset of the data (e.g. one month) and do some computation/transformation with Spark on it.

I have written the following code to select the data:

```scala
val builder = MongodbConfigBuilder(Map(Host -> List("<host>:<port>"), Database -> "<database>", Collection -> "<collection>", Credentials -> List(MongodbCredentials("<username>","<database>","<password>".toCharArray))))
val config = builder.build()

val schema = StructType(Seq(
		StructField("createdAt",TimestampType,true),
		// some more fields
))

val mongoRDD = sqlContext.fromMongoDB(readConfig, Some(schema))
mongoRDD.registerTempTable("mytable")
val resultFromMongo = sqlContext.sql("SELECT * FROM mytable WHERE createdAt >= '2015-12-01' AND createdAt < '2016-01-01'")
```

The select is quite slow even if I change the code to only select on day. It seems to me (I am quite new to Spark), that the whole collection data is loaded and the filtering is done by Spark in memory. I would like to only select and fetch the data of the given period directly from MongoDB. 

Is it possible to achieve this with your connector?

BTW: Your connector already helped me a lot. Thank you very much for your effort!

**Comment by JSantosP** [comment]
Is it possible that, even that the schema is specified, sampling is still being used?
As you can see at 

https://github.com/Stratio/spark-mongodb/blob/0.10.1/spark-mongodb/src/main/scala/com/stratio/datasource/mongodb/MongodbRelation.scala#L60 

only schema is inferred when no other one has been provided (which is not the case). What do you think @pmadrigal ?

**Comment by pmadrigal** [comment]
You are right @JSantosP , whole collection shouldn't be loaded.

@psh-tc could you give us more info? 
Spark and connector version?

**Comment by psh-tc** [comment]
I am using the spark-mongodb_2.10-0.10.1 - Package with Spark 1.5.2.

**Comment by darroyocazorla** [comment]
Hi @psh-tc,

Unlikely, Spark cannot convert some predicates into data source filters using SQL, so in this case the whole collection is loaded. When working with timestamp, Spark only supports equal filters pushdown.

Try using DataFrame DSL, It makes the type explicit and should solve the issue:

val df = sqlContext.table("mytable")
val resultFromMongo = df.filter(df("createdAt") >= java.sql.Timestamp ...)

I hope this help

**Comment by psh-tc** [comment]
Hi there,
I've finally had some time to examine the issue further.

By setting the following in MongoDB I've had a look at the queries that are executed:
```
db.getProfilingStatus()
{ "was" : 2, "slowms" : 10 }

db.system.profile.find().pretty()
```

The following Spark code
```
val df = sqlContext.table("mytable")
val resultFromMongo1 = df.filter(df("createdAt") >= java.sql.Timestamp.valueOf("2015-11-14 10:00:00")).filter(df("createdAt") < java.sql.Timestamp.valueOf("2015-11-14 10:05:00"))
resultFromMongo1.count()

val resultFromMongo2 = sqlContext.sql("SELECT * FROM mytable WHERE createdAt >= CAST('2015-11-14 10:00:00' AS TIMESTAMP) AND createdAt < CAST('2015-11-14 10:05:00' AS TIMESTAMP)")
resultFromMongo2.count()
```
results in the same query (from the profiling information):
```
{
        "op" : "query",
        ...
        "query" : {
                "find" : "<collection>",
                "filter" : {
                        "createdAt" : {
                                "$gte" : ISODate("2015-11-14T10:00:00Z"),
                                "$lt" : ISODate("2015-11-14T10:05:00Z")
                        }
                },
                "max" : {
                        "_id" : ObjectId("<some id>")
                },
                "min" : {
                        "_id" : ObjectId("<some id>")
                }
        },
        "keysExamined" : 6757,
        "docsExamined" : 6757,
        "cursorExhausted" : true,
        ...
        "nreturned" : 0,
        "responseLength" : 20,
        "millis" : 123,
        "execStats" : {
                "stage" : "PROJECTION",
                "nReturned" : 0,
                "executionTimeMillisEstimate" : 120,
                ...
                "inputStage" : {
                        "stage" : "FETCH",
                        "filter" : {
                                "$and" : [
                                        {
                                                "createdAt" : {
                                                        "$lt" : ISODate("2015-11-14T10:05:00Z")
                                                }
                                        },
                                        {
                                                "createdAt" : {
                                                        "$gte" : ISODate("2015-11-14T10:00:00Z")
                                                }
                                        }
                                ]
                        },
                        "nReturned" : 0,
                        "executionTimeMillisEstimate" : 120,
                        "works" : 6758,
                        "...
                        "docsExamined" : 6757,
                        "alreadyHasObj" : 0,
                        "inputStage" : {
                                "stage" : "IXSCAN",
                                "nReturned" : 6757,
                                "executionTimeMillisEstimate" : 15,
                                "works" : 6758,
                                "advanced" : 6757,
                                ...
                                "keyPattern" : {
                                        "_id" : 1
                                },
                                "indexName" : "_id_",
                                ...
                        }
                }
        },
        "ts" : ISODate("2016-01-21T15:13:00.580Z"),
        ...
}
```

I found while my spark job is running lots and lots ot these entries appear within the profiling output. The queries are executed within a interval of only some milliseconds.

With both variations DataFrame and Spark SQL the count of some thousand documents took 6-7 min (there are several million documents within the collection). From my perspective this takes to long for a simple count on some thousand documents (even if the documents are fetch by Spark).

Conclusion:
--> The conversion of a Spark filter (Spark SQL or DataFrame) using timestamp type into the respective MongoDB query works correctly 

Therefore the following new questions (also as I am not an expert in reading the mongoDB query profiling output):
- Why are that many queries triggered within MongoDB? (Probably due to a cursor, but why is "cursorExhausted" : true in the mongoDB profiling output?)
- It seems that the IXSCAN happens on the _id field. I had expected it on the createdAt field? (Probably due to the fact that min, max on the _id field is added to the query by some kind of cursor or pagination?)
- Why is "nreturned" always 0 (from what I can see in the output)?

At the moment it seems to me that the cursor / pagnation funtionality triggered by the connector is related to this issue.

**Comment by psh-tc** [comment]
Hi there, any news on this issue? Could anyone reproduce this behaviour?

**Comment by psh-tc** [comment]
I guess I finally understood how the connector works for non-shared collections:

1. The whole collection is splitted into partitions via the splitVector command of MongoDB. The splitKey field (by default "_id") is used to determine the min/max borders of each partition.
2. The query is then applied to each partition (see query with min/max above)

For our use case this behaviour results in thousand of queries that are fired via mongodb (as the result of the split are thousand of partitions). But, as the query only effect data in 1 or 2 partitions, most of these queries return empty. 

Therefore this behaviour results in a very poor performance for use cases where you only want to work on a small subset of a collection.

I think the best solution would be to add min and max boundaries to the splitVector command (as also proposed in #53). This would make it possible to set the boundaries for the split command to the same values as of the query (or it would even make is unnecessary to set $lt and $gt in the query).

So the connector would the run e.g. the following command:
´´´
var result = db.runCommand({splitVector: 'mydb.mycollection', keyPattern: {createdAt: 1}, min: {createdAt: new ISODate("2015-11-14T10:00:00.000Z")}, max: {createdAt: new ISODate("2015-11-14T10:05:00.000Z")}, maxChunkSize: 1})
´´´

Something similar is implemented in:
https://github.com/mongodb/mongo-hadoop/blob/master/core/src/main/java/com/mongodb/hadoop/splitter/StandaloneMongoSplitter.java

**Comment by psh-tc** [comment]
@pmadrigal: would it be possible to add min/max for the splitVector in the next version? This issue is really a blocker for us.

I would happily contribute in adding some documentation on the partioning behaviour via the splitVector command for non-shared collections.

**Comment by pmadrigal** [comment]
@psh-tc splitVector with min/max support is available in this PR #98, so you can check it.

**Comment by psh-tc** [comment]
@pmadrigal I found one issue. Details about it in the PR.

**Comment by pmadrigal** [comment]
@psh-tc it is already fixed and merged in branch 0.11 ;)

Thanks for your feedback!

**Comment by psh-tc** [comment]
Perfect! This is a fantastic improvement for my use case. Can't wait until this is releases ;) I guess it will be in 0.11.1, correct?
